File: internal/intelligence/dashboard/guidance.go
package dashboard

import (
	"context"
	"errors"
	"log/slog"
	"net/http"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"time"

	"github.com/marmutapp/superbased-observer/internal/fsview"
	"github.com/marmutapp/superbased-observer/internal/guidance"
	"github.com/marmutapp/superbased-observer/internal/models"
	"github.com/marmutapp/superbased-observer/internal/scrub"
	"github.com/marmutapp/superbased-observer/internal/store"
)

// -----------------------------------------------------------------------------
// Agent-guidance inventory API.
//
//	GET /api/projects/guidance?root=<abs>              → inventory for one root
//	GET /api/projects/guidance/file?root=<abs>&rel=…   → one file's body
//	GET /api/projects/guidance/summary                 → per-root roll-up
//
// The inventory itself is metadata the scan loop already persisted (names,
// sizes, hashes, front-matter) — no filesystem work happens on the inventory
// request. Only the /file endpoint touches disk, and it does so through the
// SAME capped, symlink-safe reader the per-terminal project panel uses
// (internal/fsview), then scrubs the text. There is no raw mode: a guidance
// file is prose an operator may have pasted a token into, and the viewer is a
// convenience, not an export.
//
// Two containment rules, both structural:
//
//  1. `root` must be a project root the store already knows. Without this the
//     endpoint would be an arbitrary-directory reader with an fsview fig leaf.
//  2. `rel` is resolved by fsview, which rejects absolute paths, traversal and
//     symlink escapes and never echoes an absolute path in its errors. A
//     user-scope row ("~/…") anchors at the operator's home instead of the
//     project root — see guidanceFileAnchor — because that is where the file
//     actually is; without it the panel offered a link that always 404'd.
// -----------------------------------------------------------------------------

// guidanceReadCap is the per-file read cap the guidance viewer applies. The
// same 256KB fsview default the project panel uses — a guidance file far past
// that is pathological, and the caller is told it was truncated.
const guidanceReadCap = fsview.DefaultMaxReadBytes

// guidanceToolsMeasurable is THE static answer to "can Observer tell you
// whether this tool actually USED a guidance file?" — a per-tool capability
// flag, not a per-tool behaviour branch (CLAUDE.md #3: the surface renders
// from the flag, it does not switch on the tool name).
//
// true means the tool emits a capture signal Observer already ingests that
// names a guidance file, AND that signal's target is GROUNDED against a real
// transcript — because since wave 2 this flag no longer merely enables a
// sentence, it enables a per-file COUNT, and a count of zero prints as "never
// invoked". Claiming measurability for a signal whose target might be empty or
// might be a path spelled differently from the inventory turns an unverified
// join into a fabricated statement of disuse. That is worse than saying
// nothing.
//
// Only claude-code clears that bar today:
//   - `skill_invoke` from the `Skill` tool call, whose target is the skill
//     NAME (internal/adapter/claudecode/adapter.go::extractTarget, grounded
//     against a live transcript in wave 2), and
//   - `instructions_loaded` from its hook (cmd/observer/hook.go,
//     buildClaudeInstructionsLoadedEvent), whose target is the file path.
//     No other adapter produces this signal at all.
//
// DELIBERATELY NOT LISTED, pending grounding: deepseek, muse, mistral-code and
// freebuff each map a native skill call onto models.ActionSkillInvoke
// (internal/tooltax/table.go), so the ACTION exists — but their targets come
// from adapter-specific extraction that has never been checked against a real
// session, and may be a path or empty. Each becomes `true` here when a
// per-adapter fixture test pins what its skill_invoke target actually
// contains, not before. Until then those tools render "usage is not
// measurable", which is exactly true.
//
// A tool absent from this map is false: absence is the zero value and the zero
// value is the honest one.
var guidanceToolsMeasurable = map[string]bool{
	"claude-code": true,
}

// guidanceMeasurableFor projects the static table onto the tools actually
// present in one inventory, so the wire map answers exactly the questions the
// rows raise — never the whole registry.
func guidanceMeasurableFor(rows []store.GuidanceRow) map[string]bool {
	out := make(map[string]bool, len(rows))
	for _, r := range rows {
		out[r.Tool] = guidanceToolsMeasurable[r.Tool]
	}
	return out
}

// The per-row usage enum. Three values, and the third is the whole point:
// `not_measurable` is NOT a zero count, it is the absence of any capture
// signal that could name this file. A surface that collapses it into
// `never_invoked` is fabricating evidence of disuse.
const (
	guidanceUsageInvoked       = "invoked"
	guidanceUsageNeverInvoked  = "never_invoked"
	guidanceUsageNotMeasurable = "not_measurable"
)

// guidanceKindSignal is the per-KIND half of measurability: which capture
// signal, if any, can ever name a guidance file of this kind. A kind absent
// here (agent, command, rule, config) has no naming signal from ANY tool, so
// its rows are always `not_measurable` however capable the tool is.
var guidanceKindSignal = map[string]string{
	string(guidance.KindSkill):        models.ActionSkillInvoke,
	string(guidance.KindInstructions): models.ActionInstructionsLoaded,
}

// guidanceSignalEmitters is the per-TOOL half: which tools emit each signal
// WITH A GROUNDED TARGET. skill_invoke reuses guidanceToolsMeasurable, which
// is narrower than the skill_invoke taxonomy row set on purpose (see that
// table's comment); instructions_loaded comes only from Claude Code's hook
// (cmd/observer/hook.go::buildClaudeInstructionsLoadedEvent) and no adapter
// produces an equivalent.
var guidanceSignalEmitters = map[string]map[string]bool{
	models.ActionSkillInvoke:        guidanceToolsMeasurable,
	models.ActionInstructionsLoaded: {"claude-code": true},
}

// guidanceUsageMeasurable resolves both halves: can Observer observe THIS
// tool using a guidance file of THIS kind? It is a capability lookup, not a
// tool-name branch (CLAUDE.md #3/#5) — a new adapter that gains a skill_invoke
// taxonomy row becomes measurable by adding one map entry, not a case.
func guidanceUsageMeasurable(tool, kind string) bool {
	signal, ok := guidanceKindSignal[kind]
	if !ok {
		return false
	}
	return guidanceSignalEmitters[signal][tool]
}

// guidanceRowOut is one inventory row plus its joined usage facts. The
// embedded store.GuidanceRow flattens into the same JSON object it always
// produced, so the addition is purely additive for an older frontend.
type guidanceRowOut struct {
	store.GuidanceRow
	// InvokedCount / LastInvokedAt come from skill_invoke actions joined on
	// guidance.SkillKey. LoadedCount / LastLoadedAt come from
	// instructions_loaded actions joined on the file's absolute path.
	InvokedCount  int    `json:"invoked_count"`
	LastInvokedAt string `json:"last_invoked_at,omitempty"`
	LoadedCount   int    `json:"loaded_count"`
	LastLoadedAt  string `json:"last_loaded_at,omitempty"`
	// Usage is one of invoked / never_invoked / not_measurable.
	Usage string `json:"usage"`
}

// guidanceRowsWithUsage folds one project's usage aggregate onto its
// inventory rows. Both joins are lookups against pre-normalized keys, so this
// is O(rows) and does no further SQL.
//
// available=false means the usage read did not happen (timed out, or failed).
// Every row then reports `not_measurable` — the ONE honest degrade, because
// "we could not measure this" is literally what happened. Reporting
// never_invoked instead would turn a database timeout into a claim about the
// operator's skills.
func guidanceRowsWithUsage(rows []store.GuidanceRow, usage store.GuidanceUsage, available bool) []guidanceRowOut {
	out := make([]guidanceRowOut, 0, len(rows))
	for _, r := range rows {
		o := guidanceRowOut{GuidanceRow: r, Usage: guidanceUsageNotMeasurable}
		if !available {
			out = append(out, o)
			continue
		}
		if stat, ok := usage.Skills[guidance.SkillKey(r.Tool, guidance.Kind(r.Kind), r.Name)]; ok {
			o.InvokedCount = stat.Count
			o.LastInvokedAt = guidanceStamp(stat.Last)
		}
		if r.AbsPath != "" {
			if stat, ok := usage.Files[filepath.Clean(r.AbsPath)]; ok {
				o.LoadedCount = stat.Count
				o.LastLoadedAt = guidanceStamp(stat.Last)
			}
		}
		if guidanceUsageMeasurable(r.Tool, r.Kind) {
			if o.InvokedCount > 0 || o.LoadedCount > 0 {
				o.Usage = guidanceUsageInvoked
			} else {
				o.Usage = guidanceUsageNeverInvoked
			}
		}
		out = append(out, o)
	}
	return out
}

// guidanceStamp renders a usage timestamp, or the empty string for "never"
// so the omitempty tag keeps the key off the wire entirely rather than
// shipping a zero time a frontend would render as 1 January year 1.
func guidanceStamp(t time.Time) string {
	if t.IsZero() {
		return ""
	}
	return t.UTC().Format(time.RFC3339)
}

// guidanceInventoryResp is the wire shape of GET /api/projects/guidance.
type guidanceInventoryResp struct {
	Root string `json:"root"`
	// ScannedAt is the most recent last_scanned across the returned rows,
	// RFC3339. Empty means this root has never been scanned — distinct
	// from "scanned and found nothing", which returns a zero-length rows
	// array WITH a timestamp.
	ScannedAt string           `json:"scanned_at,omitempty"`
	Rows      []guidanceRowOut `json:"rows"`
	// ToolsMeasurable answers, per tool present in Rows, whether Observer
	// can observe that tool USING its guidance files. See the table above.
	ToolsMeasurable map[string]bool `json:"tools_measurable"`
	// UsageWindowDays is the lookback the per-row counts cover, so a
	// surface can say "invoked 4x in 90d" without hard-coding the number.
	UsageWindowDays int `json:"usage_window_days"`
	// UsageSince is the exact start of that window, RFC3339.
	UsageSince string `json:"usage_since,omitempty"`
	// UsageUnavailable reports that the usage read did not complete, so
	// every row's `usage` is not_measurable for that reason rather than
	// because the tool emits no signal. A surface MUST distinguish the two:
	// one is "Observer cannot see this", the other is "Observer did not
	// look this time".
	UsageUnavailable bool `json:"usage_unavailable,omitempty"`
	// UsageErrorClass is a short closed enum (timeout | query), never the
	// driver's message — the inventory route is remote-reachable and a raw
	// SQL error is host detail.
	UsageErrorClass string `json:"usage_error_class,omitempty"`
}

// guidanceUsageTimeout bounds the usage join on a dashboard GET.
//
// The dashboard server sets only ReadHeaderTimeout, so nothing upstream of
// this handler would ever cut a slow query off: a cold seek over a
// multi-gigabyte actions table would simply hang the panel with no inventory
// at all, even though the inventory itself is a cheap read that was already
// done. Three seconds is well past the warm case and well short of a user
// deciding the page is broken.
const guidanceUsageTimeout = 3 * time.Second

// Closed vocabulary for guidanceInventoryResp.UsageErrorClass.
const (
	guidanceUsageErrTimeout = "timeout"
	guidanceUsageErrQuery   = "query"
)

// guidanceUsageFn is the usage-read seam, so the fail-soft branch is testable
// without a slow database. Nil means the real store read.
type guidanceUsageFn func(ctx context.Context, root string) (store.GuidanceUsage, error)

// readGuidanceUsage runs the usage join under its own deadline and converts
// every failure into the fail-soft triple (zero usage, available=false, error
// class). It never returns an error: the inventory is the primary answer and
// must still be served.
func readGuidanceUsage(ctx context.Context, read guidanceUsageFn, root string) (store.GuidanceUsage, bool, string) {
	uctx, cancel := context.WithTimeout(ctx, guidanceUsageTimeout)
	defer cancel()

	usage, err := read(uctx, root)
	if err == nil {
		return usage, true, ""
	}
	class := guidanceUsageErrQuery
	// The CALLER going away is not a usage failure (the response is
	// discarded anyway); only our own deadline is a timeout.
	if errors.Is(err, context.DeadlineExceeded) || (uctx.Err() != nil && ctx.Err() == nil) {
		class = guidanceUsageErrTimeout
	}
	return store.GuidanceUsage{}, false, class
}

// guidanceFileResp is the wire shape of GET /api/projects/guidance/file.
type guidanceFileResp struct {
	RelPath   string `json:"rel_path"`
	Content   string `json:"content"`
	Truncated bool   `json:"truncated"`
	// Scrubbed is always true — the viewer has no raw mode.
	Scrubbed bool `json:"scrubbed"`
}

// handleProjectGuidance serves the per-root guidance inventory.
func (s *Server) handleProjectGuidance(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodGet {
		writeProjectError(w, http.StatusMethodNotAllowed, "method_not_allowed")
		return
	}
	root, ok := s.knownProjectRoot(r)
	if !ok {
		writeProjectError(w, http.StatusNotFound, "unknown_root")
		return
	}
	st := store.New(s.db())
	rows, err := st.ListGuidance(r.Context(), root, false)
	if err != nil {
		writeErr(w, err)
		return
	}
	if rows == nil {
		rows = []store.GuidanceRow{}
	}
	sortGuidanceRows(rows)

	resp := buildGuidanceInventoryResp(r.Context(), root, rows,
		func(ctx context.Context, root string) (store.GuidanceUsage, error) {
			return st.GuidanceUsage(ctx, root, time.Time{})
		})
	// One line at WARN per degraded request: a panel that quietly stops
	// reporting usage is exactly the kind of silent degrade an operator
	// should be told about once. A cancelled CALLER is not a degrade.
	if resp.UsageUnavailable && r.Context().Err() == nil && s.opts.Logger != nil {
		s.opts.Logger.Warn("dashboard: guidance usage join unavailable, serving inventory without it",
			slog.String("class", resp.UsageErrorClass))
	}
	writeJSON(w, resp)
}

// buildGuidanceInventoryResp assembles the inventory response around a
// BOUNDED, FAIL-SOFT usage read.
//
// It is a separate function from the handler for one reason: the degraded
// branch is the interesting one, and it is testable here with an injected
// failing or slow `read` instead of a genuinely slow database.
func buildGuidanceInventoryResp(
	ctx context.Context,
	root string,
	rows []store.GuidanceRow,
	read guidanceUsageFn,
) guidanceInventoryResp {
	// The usage join is a second read of the SAME database, over the biggest
	// table in it. The inventory is the primary answer and a slow or failing
	// aggregate must never take it down — so the read is bounded and its
	// failure degrades every row to not_measurable, flagged explicitly.
	// Degrading to never_invoked instead would turn a database timeout into
	// a claim about the operator's skills.
	usage, usageOK, usageErrClass := readGuidanceUsage(ctx, read, root)

	var newest time.Time
	for _, row := range rows {
		if row.LastScanned.After(newest) {
			newest = row.LastScanned
		}
	}
	var scannedAt string
	if !newest.IsZero() {
		scannedAt = newest.UTC().Format(time.RFC3339)
	}
	return guidanceInventoryResp{
		Root:             root,
		ScannedAt:        scannedAt,
		Rows:             guidanceRowsWithUsage(rows, usage, usageOK),
		ToolsMeasurable:  guidanceMeasurableFor(rows),
		UsageWindowDays:  store.GuidanceUsageWindowDays,
		UsageSince:       guidanceStamp(usage.Since),
		UsageUnavailable: !usageOK,
		UsageErrorClass:  usageErrClass,
	}
}

// handleProjectGuidanceFile serves one guidance file's scrubbed body.
//
// Classed alongside the per-terminal project-file route: it is a View-tier
// GET, and a remote-exposed caller is refused unless the owner has turned on
// [remote].allow_terminal_view — file CONTENT is at least as sensitive as
// terminal output, and the inventory (paths only) is the lighter surface a
// remote caller keeps.
func (s *Server) handleProjectGuidanceFile(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodGet {
		writeProjectError(w, http.StatusMethodNotAllowed, "method_not_allowed")
		return
	}
	// Remote gate FIRST, before the root is even resolved, so the response
	// cannot be used as a project-root oracle.
	if remoteExposedFromContext(r.Context()) && !s.allowTerminalView() {
		writeProjectError(w, http.StatusForbidden, "remote_view_disabled")
		return
	}
	root, ok := s.knownProjectRoot(r)
	if !ok {
		writeProjectError(w, http.StatusNotFound, "unknown_root")
		return
	}
	rel := r.URL.Query().Get("rel")
	if rel == "" {
		writeProjectError(w, http.StatusBadRequest, "bad_path")
		return
	}
	anchor, cleanRel, ok := guidanceFileAnchor(root, rel)
	if !ok {
		writeProjectError(w, http.StatusNotFound, "not_found")
		return
	}
	c, err := fsview.Read(r.Context(), anchor, cleanRel, guidanceReadCap)
	if err != nil {
		status, code := projectFsErr(err)
		writeProjectError(w, status, code)
		return
	}
	// A binary file is not guidance. fsview reports it with empty Data; say
	// so rather than returning an empty body that reads like an empty file.
	if c.Binary {
		writeProjectError(w, http.StatusBadRequest, "bad_path")
		return
	}
	writeJSON(w, guidanceFileResp{
		RelPath:   rel,
		Content:   scrub.New().String(c.Data),
		Truncated: c.Truncated,
		Scrubbed:  true,
	})
}

// guidanceFileAnchor resolves one requested inventory path to the (root, rel)
// pair fsview needs — the same rule internal/mcp's guidanceReadAnchor applies,
// because the panel makes BOTH scopes clickable.
//
// A user-scope row's rel_path is written "~/…" (internal/guidance's
// convention) and lives under the operator's home, not under the project; a
// project-scope row anchors at the project root. Containment is still
// fsview's: anchoring on a DIRECTORY and resolving the relative path inside it
// is what makes that containment mean something, rather than trusting a stored
// absolute path.
func guidanceFileAnchor(projectRoot, rel string) (anchor, cleanRel string, ok bool) {
	if strings.HasPrefix(rel, "~/") {
		home, err := os.UserHomeDir()
		if err != nil || home == "" {
			return "", "", false
		}
		rel = strings.TrimPrefix(rel, "~/")
		if rel == "" {
			return "", "", false
		}
		return home, rel, true
	}
	if projectRoot == "" || rel == "" {
		return "", "", false
	}
	return projectRoot, rel, true
}

// handleProjectGuidanceSummary serves the per-root roll-up: one entry per
// project root that has any guidance row, newest scan first. A bare JSON
// array — the frontend's project picker consumes it directly.
func (s *Server) handleProjectGuidanceSummary(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodGet {
		writeProjectError(w, http.StatusMethodNotAllowed, "method_not_allowed")
		return
	}
	rows, err := store.New(s.db()).GuidanceProjects(r.Context())
	if err != nil {
		writeErr(w, err)
		return
	}
	if rows == nil {
		rows = []store.GuidanceProjectSummary{}
	}
	writeJSON(w, rows)
}

// knownProjectRoot resolves the ?root= query parameter against the set of
// project roots the store knows, and reports ok=false for anything else.
//
// This is the containment rule that keeps /file from being an arbitrary-file
// reader: fsview guarantees "inside SOME root", this guarantees "inside a root
// the observer already recorded for itself".
func (s *Server) knownProjectRoot(r *http.Request) (string, bool) {
	want := filepath.Clean(r.URL.Query().Get("root"))
	if want == "" || want == "." {
		return "", false
	}
	roots, err := store.New(s.db()).ProjectRoots(r.Context())
	if err != nil {
		return "", false
	}
	for _, root := range roots {
		if filepath.Clean(root) == want {
			// The CLEANED spelling is what every caller needs: a
			// projects.root_path stored with a trailing slash would
			// otherwise be handed to ListGuidance, whose rows are keyed by
			// the cleaned root the scan pass wrote — and the panel would
			// render empty for a project that has an inventory.
			return want, true
		}
	}
	return "", false
}

// sortGuidanceRows orders an inventory the way the panel renders it: by tool,
// then kind, then relative path. Exported behaviour is the ORDER, which the
// frontend relies on for stable grouping.
func sortGuidanceRows(rows []store.GuidanceRow) {
	sort.SliceStable(rows, func(i, j int) bool {
		a, b := rows[i], rows[j]
		if a.Tool != b.Tool {
			return a.Tool < b.Tool
		}
		if a.Kind != b.Kind {
			return a.Kind < b.Kind
		}
		return a.RelPath < b.RelPath
	})
}

File: web/src/lib/guidance.ts
// Guidance-file discovery for the Projects page: per-project instructions /
// skills / agents / commands / rules / config files the AI tools themselves
// read (CLAUDE.md, .cursorrules, AGENTS.md, .claude/skills/*, etc). The Go
// side (`internal/...` guidance scanner, built in parallel) owns discovery +
// parsing; this module is a thin typed fetch layer + the pure grouping/label
// helpers the page and card need.
//
// Types mirror the API contract in the GUID-C bundle brief:
//   GET /api/projects/guidance/summary -> GuidanceSummaryRow[]
//   GET /api/projects/guidance?root=<abs> -> ProjectGuidanceResponse
//   GET /api/projects/guidance/file?root=<abs>&rel=<rel> -> GuidanceFileResponse

import { fetchJSON } from "@/lib/api";
import { toolMeta } from "@/lib/tools";

export type GuidanceKind =
  | "instructions"
  | "skill"
  | "agent"
  | "command"
  | "rule"
  | "config";

export type GuidanceScope = "project" | "user";

export type GuidanceRow = {
  project_root: string;
  tool: string;
  kind: GuidanceKind;
  scope: GuidanceScope;
  // rel_path only: the API deliberately does not serialize the absolute
  // host path (project-scope rows are relative to the project root,
  // user-scope rows are "~/"-prefixed and relative to the home directory).
  rel_path: string;
  name: string;
  description?: string;
  size_bytes: number;
  content_hash?: string;
  modified_at?: string;
  frontmatter?: Record<string, string>;
  present: boolean;
  first_seen?: string;
  last_scanned?: string;
  parse_error?: string;
  // Wave-2 usage join. invoked_* count skill_invoke actions matched on the
  // backend's SkillKey identity; loaded_* count instructions_loaded actions
  // matched on the file's absolute path. Both are scoped to this project and
  // to the response's usage_window_days.
  invoked_count?: number;
  last_invoked_at?: string;
  loaded_count?: number;
  last_loaded_at?: string;
  // usage is the honest three-way answer. `not_measurable` is NOT a zero:
  // it means no capture signal exists that could ever name this file for
  // this tool, so the UI must never render it as "never invoked".
  usage?: GuidanceUsageState;
};

export type GuidanceUsageState = "invoked" | "never_invoked" | "not_measurable";

export type GuidanceSummaryRow = {
  project_root: string;
  files: number;
  last_scanned?: string;
};

export type ProjectGuidanceResponse = {
  root: string;
  scanned_at?: string;
  rows: GuidanceRow[];
  // Keyed by tool id. `false` means Observer cannot tell whether the tool
  // actually invoked/used the file at all (never render "0 invocations" for
  // those — see usageLine below). `true` means a capture signal exists, and
  // since wave 2 each row of such a tool also carries a real count.
  //
  // Note this is a per-TOOL answer; measurability is really per (tool, kind)
  // — a tool can emit skill_invoke without emitting instructions_loaded — so
  // the authoritative per-row answer is the row's own `usage` field.
  tools_measurable: Record<string, boolean>;
  // usage_window_days is the lookback the per-row counts cover (90 today).
  usage_window_days?: number;
  usage_since?: string;
  // usage_unavailable means the backend's usage read did not complete
  // (bounded by a server-side timeout), so every row's `usage` is
  // `not_measurable` for THAT reason rather than because the tool emits no
  // signal. The two must read differently: "Observer cannot see this" vs
  // "Observer did not look this time".
  usage_unavailable?: boolean;
  usage_error_class?: "timeout" | "query";
};

export const DEFAULT_GUIDANCE_USAGE_WINDOW_DAYS = 90;

export type GuidanceFileResponse = {
  rel_path: string;
  content: string;
  truncated: boolean;
  scrubbed: boolean;
};

export const PROJECTS_GUIDANCE_SUMMARY_PATH = "/api/projects/guidance/summary";
export const PROJECTS_GUIDANCE_PATH = "/api/projects/guidance";
export const PROJECTS_GUIDANCE_FILE_PATH = "/api/projects/guidance/file";

export function fetchProjectsGuidanceSummary(
  signal?: AbortSignal,
): Promise<GuidanceSummaryRow[]> {
  return fetchJSON<GuidanceSummaryRow[]>(PROJECTS_GUIDANCE_SUMMARY_PATH, undefined, {
    signal,
  });
}

export function fetchProjectGuidance(
  root: string,
  signal?: AbortSignal,
): Promise<ProjectGuidanceResponse> {
  return fetchJSON<ProjectGuidanceResponse>(
    PROJECTS_GUIDANCE_PATH,
    { root },
    { signal },
  );
}

export function fetchGuidanceFile(
  root: string,
  rel: string,
  signal?: AbortSignal,
): Promise<GuidanceFileResponse> {
  return fetchJSON<GuidanceFileResponse>(
    PROJECTS_GUIDANCE_FILE_PATH,
    { root, rel },
    { signal },
  );
}

// ---------------------------------------------------------------- labels

export const KIND_ORDER: GuidanceKind[] = [
  "instructions",
  "skill",
  "agent",
  "command",
  "rule",
  "config",
];

const KIND_LABELS: Record<GuidanceKind, string> = {
  instructions: "Instructions",
  skill: "Skills",
  agent: "Agents",
  command: "Commands",
  rule: "Rules",
  config: "Config",
};

// formatKind renders the plural section label for a guidance kind. Falls
// back to a title-cased echo of the raw value for a kind the frontend
// doesn't know about yet (additive — a new backend kind never breaks the
// page).
export function formatKind(kind: string): string {
  return KIND_LABELS[kind as GuidanceKind] ?? titleCase(kind);
}

// formatTool reuses the shared tool registry so a guidance-file's tool
// label matches every other tool badge in the dashboard rather than
// maintaining a second name map.
export function formatTool(tool: string): string {
  return toolMeta(tool).label;
}

function titleCase(s: string): string {
  if (!s) return s;
  return s
    .split(/[-_\s]+/)
    .filter(Boolean)
    .map((w) => w[0].toUpperCase() + w.slice(1))
    .join(" ");
}

// ---------------------------------------------------------------- grouping

export type GuidanceKindGroup = {
  kind: GuidanceKind | string;
  rows: GuidanceRow[];
};

export type GuidanceToolGroup = {
  tool: string;
  kinds: GuidanceKindGroup[];
  total: number;
};

// groupGuidance is pure: rows in, grouped-by-tool-then-kind out. Callers
// decide scope splitting (e.g. filter to scope==="project" vs "user" before
// calling this) so the grouping itself carries no scope opinion.
//
// Ordering: tools alphabetically by their display label (stable regardless
// of API row order), kinds in the fixed KIND_ORDER with any unknown kind
// appended after, rows within a kind by name.
export function groupGuidance(rows: GuidanceRow[]): GuidanceToolGroup[] {
  const byTool = new Map<string, GuidanceRow[]>();
  for (const r of rows) {
    const list = byTool.get(r.tool);
    if (list) list.push(r);
    else byTool.set(r.tool, [r]);
  }

  const groups: GuidanceToolGroup[] = [];
  for (const [tool, toolRows] of byTool) {
    const byKind = new Map<string, GuidanceRow[]>();
    for (const r of toolRows) {
      const list = byKind.get(r.kind);
      if (list) list.push(r);
      else byKind.set(r.kind, [r]);
    }
    const kinds: GuidanceKindGroup[] = [];
    for (const kind of KIND_ORDER) {
      const kindRows = byKind.get(kind);
      if (kindRows) {
        kinds.push({ kind, rows: sortRows(kindRows) });
        byKind.delete(kind);
      }
    }
    // Any kind the fixed order doesn't know about yet — appended, sorted
    // alphabetically, never dropped.
    for (const kind of [...byKind.keys()].sort()) {
      kinds.push({ kind, rows: sortRows(byKind.get(kind)!) });
    }
    groups.push({ tool, kinds, total: toolRows.length });
  }

  groups.sort((a, b) => formatTool(a.tool).localeCompare(formatTool(b.tool)));
  return groups;
}

function sortRows(rows: GuidanceRow[]): GuidanceRow[] {
  return [...rows].sort((a, b) => a.name.localeCompare(b.name));
}

// usageLine returns the per-tool honesty line: never a fabricated
// "0 invocations" count. `measurable` is `tools_measurable[tool]` from the
// API response (undefined treated as false — the conservative, honest
// default when the backend hasn't reported on a tool at all).
export function usageLine(
  tool: string,
  measurable: boolean | undefined,
  windowDays: number = DEFAULT_GUIDANCE_USAGE_WINDOW_DAYS,
): string {
  if (!measurable) {
    return `Usage is not measurable for ${formatTool(tool)} yet`;
  }
  return `Usage tracked over the last ${windowDays}d`;
}

// ---------------------------------------------------------------- usage

// GuidanceUsagePill is the structured per-row usage fact the card renders.
// It carries no formatted time on purpose: relative-time formatting lives in
// lib/format, and keeping this pure means the "what do we actually know"
// decision is testable and has exactly one home.
//
// A row whose `usage` is `not_measurable` yields null — the tool-group header
// already says so, and a per-row pill there would be noise pretending to be
// data.
export type GuidanceUsagePill = {
  // "invoked" for skills (a skill_invoke action), "loaded" for always-on
  // instruction files (an instructions_loaded action). Different verbs
  // because they are genuinely different events.
  verb: "invoked" | "loaded";
  count: number;
  last?: string;
  never: boolean;
};

export function guidanceUsagePill(row: GuidanceRow): GuidanceUsagePill | null {
  if (!row.usage || row.usage === "not_measurable") return null;
  // A row can in principle carry both counts (a file that is both loaded and
  // invoked); the kind decides which event the pill is about.
  const loaded = row.kind === "instructions";
  const count = (loaded ? row.loaded_count : row.invoked_count) ?? 0;
  const last = loaded ? row.last_loaded_at : row.last_invoked_at;
  return {
    verb: loaded ? "loaded" : "invoked",
    count,
    last: count > 0 ? last : undefined,
    never: row.usage === "never_invoked" || count === 0,
  };
}

// GuidanceUsageSummary counts only what is MEASURABLE. `measurable` is the
// denominator a summary line may honestly use: counting a cursor rule (which
// no capture signal can name) as "not invoked" would turn an unmeasured file
// into an unused one.
export type GuidanceUsageSummary = {
  measurable: number;
  invoked: number;
};

export function summarizeGuidanceUsage(
  rows: GuidanceRow[],
  kind?: GuidanceKind,
): GuidanceUsageSummary {
  let measurable = 0;
  let invoked = 0;
  for (const r of rows) {
    if (kind && r.kind !== kind) continue;
    if (!r.usage || r.usage === "not_measurable") continue;
    measurable++;
    if (r.usage === "invoked") invoked++;
  }
  return { measurable, invoked };
}

// formatGuidanceUsageSummary renders the Projects-panel headline, or null
// when there is nothing measurable to summarise (in which case the panel says
// nothing rather than "0 of 0").
export function formatGuidanceUsageSummary(
  summary: GuidanceUsageSummary,
  windowDays: number = DEFAULT_GUIDANCE_USAGE_WINDOW_DAYS,
  noun = "skills",
): string | null {
  if (summary.measurable === 0) return null;
  return `${summary.invoked} of ${summary.measurable} ${noun} invoked in ${windowDays}d`;
}

// formatUsageUnavailable is the copy for a degraded usage read. It says what
// happened, not what the data shows, because in this state there is no data
// to show — and it must never be confused with a real "never invoked".
export function formatUsageUnavailable(
  errorClass: ProjectGuidanceResponse["usage_error_class"],
): string {
  return errorClass === "timeout"
    ? "Usage counts unavailable - the lookup timed out. The inventory below is complete."
    : "Usage counts unavailable - the lookup failed. The inventory below is complete.";
}
